include/robot_interfaces/robot_backend.hpp

```diff
@@ -26,12 +26,113 @@
 
 namespace robot_interfaces
 {
+/**
+ * @brief Timer to measure code execution time with "checkpoints"
+ *
+ * This timer is meant to be used for measuring execution time of a loop.  It
+ * measures time between calls of the `start` method, so by calling this at the
+ * beginning of the loop, you get the execution time of the full iteration.
+ * Further, you can define "checkpoints" within the loop to measure time of
+ * separate steps in the loop.  Call the `checkpoint` method after the code that
+ * is associated with it.  For each checkpoint, the time elapsed since the last
+ * checkpoint is measured (`start` counts as a checkpoint in this regard).
+ *
+ * Example:
+ *
+ * ~~~{.cpp}
+ * CheckpointTimer<3> timer;
+ *
+ * while (condition)
+ * {
+ *     timer.start();
+ *
+ *     initialize()
+ *     timer.checkpoint("initialize");
+ *
+ *     do();
+ *     some();
+ *     stuff();
+ *     timer.checkpoint("do some stuff");
+ *
+ *     write_log();
+ *     timer.checkpoint("logging");
+ * }
+ * ~~~
+ *
+ * @tparam NUM_CHECKPOINTS Number of checkpoints.
+ * @tparam ENABLED Set to false, to disable timer.  Method calls will have no
+ * effect (and should hopefully be optimized away by the compiler).
+ */
+template <size_t NUM_CHECKPOINTS, bool ENABLED = true>
+class CheckpointTimer
+{
+public:
+    //! @brief Start timer iteration.
+    void start()
+    {
+        if (ENABLED)
+        {
+            timers_[0].tac_tic();
+            current_checkpoint = 1;
+            timers_[current_checkpoint].tic();
+        }
+    }
+
+    /**
+     * @brief Set checkpoint for time measurement.
+     *
+     * Measures time from the last call of start() or checkpoint() until this
+     * call.  The given name is used when printing the results.
+     *
+     * @param checkpoint_name Name of the checkpoint (used for printing results)
+     */
+    void checkpoint(const std::string &checkpoint_name)
```

Comment:
isn't there a problem here if i don't call the checkpoints always in the same order? i.e. if all the names have been set, and the next time the user calls this function in a different order, it should throw an error i guess.

so probably a check of checkpoint_name_[current_checkpoint] == checkpoint_name would be necessary.

Reply:
Do you think this is a realistic scenario? I would like to keep the timing code as compact as possible to not delay the loop too much, as this is affecting the "Total" measurement.

Reply:
Hm, something that might actually happen is a checkpoint that is accidentally put inside an `if` and thus is not executed in every iteration...